### ISSUE TITLE
remove console=ttyS0 on metal

### ISSUE THREAD

**Issue by cgwalters** [coreos/fedora-coreos-tracker#567]
Moving this from https://bugzilla.redhat.com/show_bug.cgi?id=1839923

I think we should likely remove the `console=ttyS0` we're injecting on bare metal by default.  On real hardware we don't expect anything connected to the serial ports by default, and attempting to write to them can greatly slow down the boot.  A web search for "linux console serial slow" turns up things like https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=763601;msg=17

Today for the Live ISO we are already doing this, but the bare metal image still enables it by default.  One can work around this by using the new [installer options to remove kargs](https://github.com/coreos/coreos-installer/pull/268) at least.

Anyone who wants to turn it on can of course.

**Comment by dustymabe** [comment]
> On real hardware we don't expect anything connected to the serial ports by default

I'm not sure that's true. In every enterprise environment I've worked in there were always either dedicated console servers attached to serial ports OR out of band management (aka lights out management) provided by the hardware that gives you access to the serial console. If your server crashes or becomes unrepsonsive the output of the serial console is one of the last chances you have to find some clues.

I think I'd much prefer to document in our FAQ that this problem can exist and show people how to change it if they are having trouble.

**Comment by dustymabe** [comment]
> I think I'd much prefer to document in our FAQ that this problem can exist and show people how to change it if they are having trouble.

Though I'm definitely interested in other thoughts and opinions here. Should we bring it up during the next community meeting?

**Comment by jlebon** [comment]
Another approach which I've thought about a bit when working on https://github.com/coreos/fedora-coreos-tracker/issues/533 is to go the whole way and remove *all* the `console` kargs in the metal and ISO artifacts. The default behaviour of the kernel in that case is to simply find the first device that can act as a console. So e.g. if there's no VGA card, and ttyS1 is active, it'll automatically select that one. This also fixes the non-obvious behaviour that the *first* `console` argument wins "per type" (see https://github.com/coreos/fedora-coreos-tracker/issues/533#issuecomment-644319795), which makes appending the `console` arg you want more awkward than it should be (i.e. you also need to delete the baked-in karg).

**Comment by cgwalters** [comment]
> I'm not sure that's true. In every enterprise environment I've worked in there were always either dedicated console servers attached to serial ports OR out of band management (aka lights out management) provided by the hardware that gives you access to the serial console. 

Right; I guess this boils down to: is the "default" having a LOM attached or not?  One or the other case will need to configure things.

> Another approach which I've thought about a bit when working on #533 is to go the whole way and remove all the console kargs in the metal and ISO artifacts. The default behaviour of the kernel in that case is to simply find the first device that can act as a console. So e.g. if there's no VGA card, and ttyS1 is active, it'll automatically select that one. 

Hmm...how is "active" detected for serial consoles?  It sounds like the desired behavior in a LOM scenario is that even if a VGA card is present, output still goes to the serial console too.

**Comment by jlebon** [comment]
> It sounds like the desired behavior in a LOM scenario is that even if a VGA card is present, output still goes to the serial console too.

I'm not sure if that's possible in the generic case since different LOM systems could be using different port numbers, IIUC?

> Hmm...how is "active" detected for serial consoles?

Here's what [the docs](https://www.kernel.org/doc/html/latest/admin-guide/serial-console.html) say:

> If no console device is specified, the first device found capable of acting as a system console will be used. At this time, the system first looks for a VGA card and then for a serial port. So if you don’t have a VGA card in your system the first serial port will automatically become the console.

So yeah, I guess it'd just be COM1, which wouldn't help if LOM is in COM2.

Hmm, was testing the fallback behaviour now with `cosa run -c -- -vga none` and removing all the `console` args, but it doesn't seem to work.

**Comment by dustymabe** [comment]
> > I'm not sure that's true. In every enterprise environment I've worked in there were always either dedicated console servers attached to serial ports OR out of band management (aka lights out management) provided by the hardware that gives you access to the serial console.
> 
> Right; I guess this boils down to: is the "default" having a LOM attached or not? One or the other case will need to configure things.

Yeah. It's definitely a question of sane defaults. To me we would need to answer questions like?

- How much does the 'slow serial' problem affect actual users today?
- What do other OS platforms do when configuring the serial console?
- If we were to change our current settings would that be worse than what the current defaults are?
    - This gets into your question of the default of having a serial console attached or not.

So we have to dig in to how many people expect and depend on the current default behavior of the serial console getting kernel message output vs how many people experience slow booting systems because we default to outputting on the serial console? In which of these cases is it OK to document the current behavior and the workaround versus changing the default.

> 
> > Another approach which I've thought about a bit when working on #533 is to go the whole way and remove all the console kargs in the metal and ISO artifacts. The default behaviour of the kernel in that case is to simply find the first device that can act as a console. So e.g. if there's no VGA card, and ttyS1 is active, it'll automatically select that one.
> 
> Hmm...how is "active" detected for serial consoles? It sounds like the desired behavior in a LOM scenario is that even if a VGA card is present, output still goes to the serial console too.

Yeah. I speak for only one person, but my prefereed LOM scenario is to have serial output even if a VGA is present. I actually much prefer serial console to VGA in any cases where I'm in a non-graphical environment (copy/paste anyone?).

**Comment by cgwalters** [comment]
One thing I just found out the hard way while playing with a FSBCOS install that failed in the initramfs is that on this Lenovo T590 laptop (that doesn't have a serial device), systemd will fail to start emergency mode because it's trying to access the non-existent serial console or something.

It works to mount the `boot` partition after running `coreos-installer` and remove all of the `console` args.

**Comment by jamescassell** [comment]
> One thing I just found out the hard way while playing with a FSBCOS install that failed in the initramfs is that on this Lenovo T590 laptop (that doesn't have a serial device), systemd will fail to start emergency mode because it's trying to access the non-existent serial console or something.
> 
> It works to mount the `boot` partition after running `coreos-installer` and remove all of the `console` args.

Sounds like we need to delete the entry if there's no serial port attached, to avoid such a problem...

**Comment by jlebon** [comment]
> One thing I just found out the hard way while playing with a FSBCOS install that failed in the initramfs is that on this Lenovo T590 laptop (that doesn't have a serial device), systemd will fail to start emergency mode because it's trying to access the non-existent serial console or something.

Yeah, I was bit by this as well when installing FCOS on my local server. I think it's because `console=ttyS0` is last and so that's what gets `/dev/console` and what systemd tries to use. See also [the docs](https://docs.fedoraproject.org/en-US/fedora-coreos/emergency-shell/) we have about this.

> Sounds like we need to delete the entry if there's no serial port attached, to avoid such a problem...

One thing I [mentioned](https://github.com/coreos/fedora-coreos-tracker/issues/533#issuecomment-618024862) was the possibility of matching the state of `console` kargs of the install boot in the installed system. (So e.g. if the user booted with `console=ttyS1`, we match that; if there are no `console=` kargs at all, then we have none either.)

Clearly that doesn't cover all use cases since you could be installing differently than how you plan to run it, but it seems like a better default heuristic overall (that you of course should be able to override). OTOH, magic handling like this can also make things more confusing.

**Comment by spamcop** [comment]
One thing: you cannot assume console=ttyS0 is correct serial console. It also uses default value 9600n8 (=slow). Depending on server, correct value can be for example console=ttyS1 or console=ttyS0,115200 or there could be no serial console at all. Its better to leave it up to user to specify that or forward what he already specified using first boot as jlebon wrote.

https://www.kernel.org/doc/html/latest/admin-guide/serial-console.html

**Comment by dustymabe** [comment]
As far as default speed value goes, our current default is `console=tty0 console=ttyS0,115200n8`, so it's at least not `9600n8` (slow).

**Comment by dustymabe** [comment]
We discussed this in the [meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/PNH6LYT65J2XXZJ3G4EPXAB5WOZJZXOY/). While we didn't get unanimous agreement the general consensus was: 

```
13:17:43     dustymabe | #agreed We'll drop all console= kargs on new bare-metal installs, and
                       | implement a platform-specific default on other platforms.  Upgrades of
                       | existing machines won't be affected.  We'll document how to configure
                       | alternate consoles.

```

The platform specific defaults on other platforms plays into https://github.com/coreos/fedora-coreos-tracker/issues/110 I believe.

**Comment by ohadlevy** [comment]
I hit this issue where ttyS0 was locking up my system, ttyS1 worked just fine, any chance we can drop it first and fix a per platform defaults in a follow up release? thanks.

**Comment by jlebon** [comment]
> I hit this issue where ttyS0 was locking up my system, ttyS1 worked just fine, any chance we can drop it first and fix a per platform defaults in a follow up release? thanks.

Yeah, we still need to action this. Note for the time being, you can modify the kargs at install time using `coreos-installer install --append-karg/--delete-karg`.

**Comment by cgwalters** [comment]
> I hit this issue where ttyS0 was locking up my system,

Yes, this was also part of a big downstream OpenShift/RHCOS customer issue.  One middle ground here might be to remove the `console=ttyS0` karg after a successful installation (e.g. after Ignition has run?)

**Comment by bgilbert** [comment]
> One middle ground here might be to remove the console=ttyS0 karg after a successful installation (e.g. after Ignition has run?)

That doesn't really help the UX issue though.  The main situation where people need to interact with the console is when Ignition has failed.

**Comment by cgwalters** [comment]
> That doesn't really help the UX issue though. The main situation where people need to interact with the console is when Ignition has failed.

Right, I said remove it *after* Ignition has completed successfully.  Or am I misunderstanding you?

**Comment by cgwalters** [comment]
A bunch of console links:

- https://lwn.net/Articles/705938/
- https://lwn.net/Articles/800946/
- https://lwn.net/Articles/780556/

One thing that came up there is that still today the kernel doesn't distinguish well between "a driver hit an unexpected timeout" and "the machine is about to die".

Related to that, one recommendation in this customer case was "set `kernel.printk` to `4 4 1 7`" to ignore debug/informational messages.  This one needs analysis - what messages would we lose etc.?

**Comment by bgilbert** [comment]
> Right, I said remove it after Ignition has completed successfully. Or am I misunderstanding you?

Yeah, I'm arguing that doesn't go far enough.  When users hit an Ignition failure and try to debug it, they discover that output is being sent to a port they don't use, don't know about, and may not even have on their machine.  Per https://github.com/coreos/fedora-coreos-tracker/issues/567#issuecomment-662632315, we should just drop `console=ttyS0` by default.

**Comment by cgwalters** [comment]
> Per #567 (comment), we should just drop console=ttyS0 by default.

OK.  For users who want to keep the serial (LOM users?) and are booting from the ISO, it seems to me they would need to use `coreos-installer iso kargs modify --append=console=ttyS0,115200n8` before booting the ISO or so, right?

I find it really hard to understand the "blast radius" of this across the platforms.  I guess the main concern here would be metal and probably vSphere.

**Comment by pamoedom** [comment]
My 2 cents:

In our case (Installer QE), due to our external provider configuration, we need to modify the kernel arguments from "console=ttyS0,115200n8" to "console=ttyS1,115200n8" in order to properly redirect the console output, having the default value in place prevents us from using a simply MachineConfig (MC) solution like the following:

~~~
$ cat << EOF > 99-openshift-machineconfig-master-kargs.yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 99-openshift-machineconfig-master-kargs
spec:
  kernelArguments:
    - 'console=ttyS0,115200n8'
EOF
~~~

In order to workaround this, we need to push via iPXE a custom ignition hook that changes the kernel arguments on the following manner:

~~~
#!ipxe

kernel $kernel initrd=main console=tty0 console=ttyS1,115200n8 coreos.inst.install_dev=/dev/sda coreos.inst.ignition_url=http://<IP>:8000/rhcos/ignitions/$cluster/master.ign coreos.live.rootfs_url=$rootfs ignition.config.url=http://<IP>:8000/rhcos/ignitions/$cluster/master-console-hook.ign ignition.firstboot ignition.platform.id=metal
initrd --name main $initramfs
boot
~~~

OR, when using TPM encryption, we use live iPXE mode to boot the console hook in memory to avoid tampering with the storage:

~~~
#!ipxe

kernel $kernel initrd=main console=tty0 console=ttyS1,115200n8 coreos.live.rootfs_url=$rootfs ignition.config.url=http://<IP>:8000/rhcos/ignitions/$cluster/master-console-hook.ign ignition.firstboot ignition.platform.id=metal
initrd --name main $initramfs
boot
~~~

Here an example of the console hook if needed:

~~~
# cat /var/www/html/rhcos/ignitions/$cluster/master-console-hook.ign
{"ignition":{"version":"3.1.0"},"systemd":{"units":[{"contents":"[Unit]\nDescription=Run after install\nAfter=coreos-installer.service\nBefore=coreos-installer.target\n\n[Service]\nType=oneshot\nExecStart=/usr/bin/coreos-installer install /dev/sda --delete-karg console=ttyS0,115200n8 --append-karg console=ttyS1,115200n8 --ignition-url http://<IP>:8000/rhcos/ignitions/$cluster/master.ign --insecure-ignition\n\n[Install]\nRequiredBy=coreos-installer.target\n","enabled":true,"name":"post-install-hook.service"}]}}
~~~

In summary, if we remove the default "console=ttyS0,115200n8", we can avoid using ignition hooks and simply push master/worker MCs with the kernel argument addition right? to me sounds like a good plan.

Best Regards.

**Comment by EmmanuelKasper** [comment]
@pamoedom 
could it be that the need for removing ttyS0 simply arise from systemd starting a getty only on the *first* found serial console ? 
according to http://0pointer.de/blog/projects/serial-console.html

> However, sometimes there's the need to manually configure a serial getty, for example, if more than one serial login prompt is needed or the kernel console should be redirected to a different terminal than the login prompt. 

So maybe you could simply add a MachineConfig to create  a getty for ttyS1, and you would get a serial console (maybe not owning /dev/console) on ttyS1 ?

**Comment by EmmanuelKasper** [comment]
OK please ignore my comment above: according to https://www.kernel.org/doc/Documentation/admin-guide/serial-console.rst

> Note that you can only define one console per device type (serial, video).

**Comment by EmmanuelKasper** [comment]
Another use case where the presence of ttyS0 is problematic, is when you hook a serial console over IPMI, which defaults to ttyS1 on many systems.
As only one serial console is possible at the time (see link above), you need to modify the kargs manually on boot, or via a custom Ignition config, which although documented, is not trivial if you're not a CoreOs specialist.

**Comment by jlebon** [comment]
Re-read the meeting notes for when [this](https://github.com/coreos/fedora-coreos-tracker/issues/567#issuecomment-662632315) was discussed. It seems like there's a lot of hesitation because it's not clear of the proportion of systems relying on this vs those being hindered by it.

Should we have someone/a few folks just look at what the most prevalent setups are? E.g. let's look at Dell, HP, Intel, Cisco, etc... and see what their LOM systems expect.

> As only one serial console is possible at the time (see link above), you need to modify the kargs manually on boot

To clarify, I guess this is something you probably have to do for most OSes if you're expecting ttyS1? (The adding ttyS1, not the removing ttyS0 part.)

**Comment by EmmanuelKasper** [comment]
My expectation would be no serial console set at all by default, so I can easily add either ttyS0 or ttyS1 if needed. Maybe the default ttyS0 improved the life of some users, but in the end it deteriorates the experiences for other (packet metal users, people with 
IPMI Serial over LAN, and those noticing a slower boot on their VGA console because the kernel is poking a serial port or terminal which is not existing) So IMHO removing console=ttyS0 on metal is a step in the right direction.

**Comment by bgilbert** [comment]
[Reported case](https://bugzilla.redhat.com/show_bug.cgi?id=1985030) where a machine would not boot with `console=ttyS0,115200`.

**Comment by pcfe** [comment]
My expectation would be that if and only if I add a `console=ttyS…` statement when I PXE boot the rhcos installer it makes in into the kernel cmdline of the installed system.
That is the way RHEL installed via anaconda has been operating for as long as I can remember and that ensures that users of rhcos do not have a potentially negative experience when they use the installer out of the box.

**Comment by lucab** [comment]
@fdammeke that's a very specific dracut/NM bug for which we are waiting the upstream fix, see https://github.com/coreos/fedora-coreos-tracker/issues/943.

**Comment by alanconway** [comment]
This is  still a problem, I've been struggling to install openshift SNO on a thinkpad p1. 
First I tried removing
   console=tty0 console=ttyS0
that didn't seem to work, so next I replaced them both with 
  console=ttyS1,115200n8
that worked.
This is a pretty serious adoption problem, if somebody tries this as their first experiment instead of CRC, they are not going to be impressed.

**Comment by cgwalters** [comment]
> This is a pretty serious adoption problem, if somebody tries this as their first experiment instead of CRC, they are not going to be impressed.

I don't disagree, but bear in mind there is a flip side here where if people try provisioning OCP on their bare metal servers and suddenly they can't log in on the LOM console, those people will also not be happy.

We're all agreed we should probably require an explicit console choice, and that will likely happen.  But as stated above there's a large potential blast radius from such a change.

**Comment by alanconway** [comment]
@cgwalters Can we leave the default alone for real deployments but add a checkbox in the installer GUI somewhere?

   [ ] I am installing SNO on a laptop or desktop machine for testing purposes. (This is important so that console parameters can be set correctly during installation)

I'm not clear on the how and why of console settings - so maybe that's not the right wording - but you get the idea.

**Comment by alanconway** [comment]
If it's possible to auto-detect the correct settings that would be better, but I'm guessing that's not possible or it would have been done already :)

**Comment by alanconway** [comment]
@cgwalters intalling SNO again and hitting this issue again - is there a better alternative to interrupting the boot process and manually editing the parameters? E.g. can I edit the SNO discovery image somewhere to fix this before I start installing? I've grepped it, but the console= params aren't in plain text anywhere, Is there some coreos.XXX parameter I can set in isolinux.cfg or grub.cfg to change the console settings that are used once coreos is installed?

**Comment by jlebon** [comment]
I'm not sure how SNO is wired up, but if you're able to customize the way SNO runs `coreos-installer install`, you could add `--delete-karg console=tty0 --delete-karg console=ttyS0 --append-karg ttyS1,115200n8` to the invocation.

We've [improved](https://coreos.github.io/coreos-installer/customizing-install/) [the UX](https://coreos.github.io/coreos-installer/cmd/iso/#coreos-installer-iso-customize) around customizing installs recently, but again I'm not sure how well this meshes with the SNO flow.

**Comment by alanconway** [comment]
@jlebon To install SNO there are a couple of very simple web forms, that gives you a downloadable ISO which I put on a USB and boot my bare-metal box with it. It would be a beautiful experience if it worked :) There's no obvious place for such low-level hackery, nor should there be. The correct settings can only be determined while booting the node, so it seems like this can only be fixed in the coreos boot loader somewhere, to use working settings for TTYs based on the machine it is being booted on.

**Comment by cgwalters** [comment]
> There's no obvious place for such low-level hackery, nor should there be. The correct settings can only be determined while booting the node, so it seems like this can only be fixed in the coreos boot loader somewhere, to use working settings for TTYs based on the machine it is being booted on.

It's possible that we could move detection logic into the initramfs, and do hardware probing there, and dynamically ask the kernel to start emitting output or not.  (IOW, this wouldn't be directly controlled by kernel arguments)

My instinct says that with such an approach we could potentially "do the right thing" in at least half of cases.  But I suspect there's a fairly long tail here, and the likely consequence would be that we stop emitting serial console output on some bare metal hardware types that do have IPMI.

**Comment by cgwalters** [comment]
Short term, I think SNO should gain a configuration option for this.

**Comment by alanconway** [comment]
@cgwalters config option makes sense as interim solution, should be visible on the SNO installation UI with some comprehensible text "are you experimenting on a laptop or desktop?" or whatever is an accurate description of when this is needed.

**Comment by romfreiman** [comment]
@alanconway what sno ui u have in mind? Assisted installer on cloud.redhat.com?

**Comment by alanconway** [comment]
Yup: https://console.redhat.com/openshift/assisted-installer/clusters/~new

Actually now that I look at it: I'm using SNO as a "local" cluster, like I would use CRC, but I'm installing it from the "Data Center" tab. Maybe the way to resolve this is to have a SNO option under the "Local" tab for folks like me, where the TTY settings are for a laptop or desktop. Then the SNO option in the "Data Center" tab can be left as-is for headless hardware.

**Comment by lucab** [comment]
Cross-referencing another report from the user forum: https://discussion.fedoraproject.org/t/very-slowly-booting-fedora-coreos-on-intel-i9-hetzner-box/38670.

**Comment by dustymabe** [comment]
I suspect we'll probably send out a coreos-status note for this? Do we have any opinions on if a "notice period" is required?

**Comment by bgilbert** [comment]
I think we should probably have a notice period, yeah.

**Comment by bgilbert** [comment]
We [discussed this](https://meetbot.fedoraproject.org/teams/fedora_coreos_meeting/fedora_coreos_meeting.2022-06-08-16.30.log.html) at today's meeting.

```
  * AGREED: We will make the serial console changes in `next` two weeks
    after the coreos-status announcement, and in `testing` two months
    after `next`.  (bgilbert, 17:18:53)
```

**Comment by bgilbert** [comment]
The current design includes one regression: the GRUB menu on bare metal will not be directed to the serial console and there's no official way to reconfigure it.  Should we consider Butane sugar similar to https://github.com/coreos/butane/pull/339 for configuring the GRUB console?

**Comment by bgilbert** [comment]
If we wanted to be really complete, we could add `coreos-installer install --append-grub-cmd` too.  That seems less important, since there isn't much reason to use the GRUB menu immediately after installation.

**Comment by dustymabe** [comment]
> The current design includes one regression: the GRUB menu on bare metal will not be directed to the serial console and there's no official way to reconfigure it.

Is the current (legacy) behavior with respect to GRUB+serial console problematic?

I just tried in a VM and can verify that I get GRUB on both serial and VGA with the current `stable` after an install from ISO (all defaults). I'm using a VM (`virt-install`) so maybe there is some special handling there that isn't true on real bare metal hardware.

This is a long way of me asking: "should we only change the kernel console options and leave the serial console configuration in GRUB for bare metal?"

**Comment by bgilbert** [comment]
Hmm.  It seems unsound to assume that the first serial port isn't attached to some non-console device, and that the only possible consoles are VGA and the first serial port.

Also, we've had reports of systems that boot slowly, or fail to boot, if Linux serial console is enabled.  I think those particular cases were fixed solely by dropping `console=ttyS0`, but I'm not confident that GRUB serial console will never cause similar issues.

**Comment by ktdreyer** [comment]
I installed SNO today over PXE. Upon booting up, the kernel console is very slow. It's taking almost an hour to boot the node because it prints each kernel line slowly, like multiple seconds between each line. In CentOS 8 or other OSes it boots very fast.

Here are my PXE settings I used to install SNO. `vmlinuz`, `initrd.img`, `ignition.img`, and `rootfs.img` all came from my SNO discovery iso I downloaded from RH's assisted installer website.

```
LABEL openshift-sno
KERNEL ocp410/vmlinuz
APPEND initrd=ocp410/initrd.img,ocp410/ignition.img random.trust_cpu=on ignition.firstboot ignition.platform.id=metal coreos.live.rootfs_url=http://server.ktdreyer.com/openshift/rootfs.img console=ttyS1 coreos.inst.persistent-kargs="console=ttyS1"
```

Compare with a bog-standard CentOS 8 config where it boots normally, both into Anaconda and the resulting OS:
```
LABEL CentOS-Stream-8
KERNEL el8/x86_64/vmlinuz
APPEND initrd=el8/x86_64/initrd.img inst.ks=http://server.ktdreyer.com/el8/x86_64/ks.cfg console=ttyS1
```

What can I set in order to make it boot as fast as CentOS?

**Comment by bgilbert** [comment]
Let's keep this ticket focused on Fedora CoreOS, please.  Support questions for RHCOS, SNO, and the Assisted Installer are best pursued via other channels.

**Comment by bgilbert** [comment]
@jlebon convinced me that `coreos-installer` sugar, including for `iso customize`, makes more sense than Butane sugar here.  I'm going to try writing up a design doc.

**Comment by newkit** [comment]
I am seeing this long boot times related to the console settings  on Lenovo x3850 X6 servers. Would be great to see this in production release soon since the workaround of removing the console settings from the commandline is tedious.

**Comment by bgilbert** [comment]
No design doc, but https://github.com/coreos/coreos-installer/pull/977 is a draft PR adding `coreos-installer install --console` and corresponding `iso/pxe customize` options.  These allow specifying one or more consoles to be used by both the bootloader (GRUB) and the kernel.

**Comment by bgilbert** [comment]
[coreos-installer 0.16.0](https://github.com/coreos/coreos-installer/releases/tag/v0.16.0) includes support for `install --console` and `iso/pxe customize --dest-console`.

**Comment by bgilbert** [comment]
The change has been [announced](https://lists.fedoraproject.org/archives/list/coreos-status@lists.fedoraproject.org/message/GHLXX4MXNHUEAXQLK6BZN45IQYHRVQB4/) to [coreos-status](https://lists.fedoraproject.org/archives/list/coreos-status@lists.fedoraproject.org/).  Please see the announcement message for details.  The transition schedule is:

- `next`: week of October 3
- `testing`: week of November 28
- `stable`: week of December 12

**Comment by jcpowermac** [comment]
After reviewing all the communication I didn't see a reason for vSphere not to be included in the `platform.json`. I can for troubleshooting purposes add a serial device to a FCOS/RHCOS virtual machine(s) and send that output to either a serial concentrator or to a file on the datastore. 

What made me discover this work was an issue I am trying to investigate for OCP 4.12 on vSphere:
https://issues.redhat.com/browse/OCPBUGS-1047
where serial output is invaluable.

**Comment by dustymabe** [comment]
The fix for this went into `next` stream release `37.20221003.1.0`. Please try out the new release and report issues.

**Comment by bgilbert** [comment]
@jcpowermac Thanks for the report.  https://github.com/coreos/fedora-coreos-config/pull/2038 enables secondary serial console on VMware, and also on OpenStack and VirtualBox.

**Comment by dustymabe** [comment]
The fix for this went into `testing` stream release `37.20221127.2.0`. Please try out the new release and report issues.

**Comment by dustymabe** [comment]
The fix for this went into `stable` stream release `37.20221127.3.0`.
